### ISSUE TITLE
Type-ahead select: support key-value filtering

### ISSUE THREAD

**Issue by mcarrano** [patternfly/patternfly-design#1088]
Want to add the ability to use a type-ahead select for selecting multiple key-value pairs as shown in the examples for ACM in this doc: https://docs.google.com/document/d/11U-lJ6B7l6uGPW6A-lgteFREY874mzBhy1LxHZfNQb8/edit?usp=sharing

**Comment by mcarrano** [comment]
Here is an initial proposal for how this could work by enhancing existing search input behavior.

![Type-ahead-attribute-value.png](https://images.zenhubusercontent.com/5ce816524e6c594dc1b0df4d/04992b08-57b8-4f5f-ae8f-0817cfe08429)

**Comment by imjoyjean** [comment]
wondering what this would look like at scale? If there were 5 key-value pairs, would the search bar expand to be bigger or would the extra key-values that don't fit be truncated?

**Comment by mmenestr** [comment]
Yeah, good question Joy! When you look at how the github key:value search works, it basically shifts everything to the left when you type more things than can fit in the bar, so that you always see the most recent thing you typed in. Maybe we could do that? Cause I definitely don't think the search bar should expand to be bigger as the user types haha

**Comment by MariSvirik** [comment]
In my opinion, enlarging the search field shouldn't be an option. That's what Satellite does in some places. 
The truncated text would be problematic as it limits legibility. Also, I'm not a big fan of scrolling right and left to read the entire query. Unfortunately, I can't think of a better solution (maybe just chips under the search bar - that would accommodate more pairs)

Is there any reasoning behind why we use the group names (attributes, values) in the list?

**Comment by mcarrano** [comment]
@mmenestr @imjoyjean @MariSvirik Good question about how this should scale. The expectation is that we would build this off of the existing type-ahead select multiple component. That is currently implemented so it can hide items after a certain threshold is reached and wrap items once the content no longer fits the width of the container. I recorded a screen capture video to show what that looks like. You can also play with it yourself here: https://www.patternfly.org/v4/components/select#multiple


https://user-images.githubusercontent.com/6619860/131565205-2da78258-3eed-4f7e-b42b-1f2d56704b78.mov

If we went with that approach, I'd need to investigate more how that plays in the toolbar since the width is held constant but the height changes.

Regarding @MariSvirik 's question about group names, I don't feel strongly about whether they add value. @imjoyjean what do you think? I was mainly playing off your ACM example.

**Comment by Rohoover** [comment]
Just wondering if this allows for search strings?... subtracting/filtering out vs. adding chips? 

Like I want all things that start with the Letter A, and have a green status but not with the label dev.

**Comment by mcarrano** [comment]
@Rohoover No, this component would not solve for that use case. This would only address the case of wanting to filter on multiple labels/tags. There is a separate issue that I am working on for a general query builder here: https://github.com/patternfly/patternfly-design/issues/972  It may be that these can be combined into a single solution, but I have not been able to find that sweet spot yet. Welcome ideas.

**Comment by mcarrano** [comment]
I did some additional work on this to address the overflow (scalability) case that was raised above. I don't think there is a perfect solution to this, but you can see some ideas here: https://marvelapp.com/prototype/15dbe4c9 Comments welcome. I will also be presenting this at next week's PF design share.

**Comment by mcarrano** [comment]
I have made some updates based on feedback from the Sep 8 design review and comments on the [Marvel deck](https://marvelapp.com/prototype/15dbe4c9/screen/81845426) and opened https://github.com/patternfly/patternfly-react/issues/6290 to begin development in our upcoming sprint. Let me know if you have further feedback. 

Note that this is only one of several issues we will be addressing to provide a more robust solution for searching and filtering. You can follow other work here: https://github.com/patternfly/patternfly-design/issues/1089

**Comment by mcarrano** [comment]
@mmenestr This has been implemented as a demo here: https://www.patternfly.org/v4/components/text-input-group/react-demos  Not sure how or where this should be documented for designers. Should it be part of the Filter guidelines? I don't think there is any impact of the Sketch library, but please review @doruskova 

Please link any related issues top the epic. Thanks.

**Comment by doruskova** [comment]
@mcarrano You're right! It doesn't have an impact on the Sketch library. Everything should be there :)

**Comment by mmenestr** [comment]
@mcarrano created this design doc issue: https://app.zenhub.com/workspaces/design-doc-workspace-5f08c64d6a6ef2001d56a458/issues/patternfly/patternfly-org/2759

Left a comment for you in that issue for more clarification...
